src/hotspot/share/utilities/nativeCallStack.hpp

```diff
@@ -123,9 +124,21 @@ class NativeCallStack : public StackObj {
     return (unsigned int)hash;
   }
 
+  void print_frame(outputStream* out, address pc) const;
   void print_on(outputStream* out) const;
 };
 
 #define FAKE_CALLSTACK NativeCallStack(NativeCallStack::FakeMarker::its_fake)
 
+// When printing many instances of NativeCallStack that may share many PC addresses,
+// a NativeCallStackPrinter improves performance by caching printed frames by address.
+class NativeCallStackPrinter {
+  struct Entry { char text[1024]; };
```

Comment:
1024 is big enough? Would 512 suffice?

Not that important if we were creating/destroying it as needed, but right now it is always there...

Reply:
Frame strings can get very lengthy. Think templatized function names, or long file names. In theory, 1024 can be not enough. 

I first wanted to do this via strdup, but that would mean that we use twice as many mallocs, copy the strings around unnecessarily, and need to manually deallocate after reporting. So a fixed-size entry is a compromise for simplicity.

Note that this is only used during detail reports. Those are memory-hungry anyway, since we snapshot the whole state.